src/binary_agreement/binary_agreement.rs

```diff
@@ -38,6 +40,103 @@ impl<N> From<bool> for CoinState<N> {
     }
 }
 
+/// Binary Agreeement messages received from other nodes for a particular Binary Agreement epoch.
+#[derive(Debug)]
+struct ReceivedMessages {
+    /// Received `BVal` messages.
+    bval: BoolSet,
+    /// Received `Aux` messages.
+    aux: BoolSet,
+    /// Received `Conf` message.
+    conf: Option<BoolSet>,
+    /// Received `Term` message.
+    term: Option<bool>,
+    /// Received `Coin` message, namely its `SignatureShare`.
+    coin: Option<SignatureShare>,
+}
+
+impl ReceivedMessages {
+    fn new() -> Self {
+        ReceivedMessages {
+            bval: bool_set::NONE,
+            aux: bool_set::NONE,
+            conf: None,
+            term: None,
+            coin: None,
+        }
+    }
+
+    /// Inserts new message content if it is accepted or returns a `FaultKind` indicating an issue
+    /// that prevented insertion of that new content.
+    fn insert(&mut self, content: MessageContent) -> Option<FaultKind> {
+        match content {
+            MessageContent::SbvBroadcast(sbv) => match sbv {
+                sbv_broadcast::Message::BVal(b) => {
+                    if !self.bval.insert(b) {
+                        return Some(FaultKind::DuplicateBVal);
+                    }
+                }
+                sbv_broadcast::Message::Aux(b) => {
+                    if !self.aux.insert(b) {
+                        return Some(FaultKind::DuplicateAux);
+                    }
+                }
+            },
+            MessageContent::Conf(bs) => {
+                if self.conf.is_none() {
+                    self.conf = Some(bs);
+                } else {
+                    return Some(FaultKind::MultipleConf);
+                }
+            }
+            MessageContent::Term(b) => {
+                if self.term.is_none() {
+                    self.term = Some(b);
+                } else {
+                    return Some(FaultKind::MultipleTerm);
+                }
+            }
+            MessageContent::Coin(msg) => {
+                if self.coin.is_none() {
+                    self.coin = Some(msg.to_sig().clone());
+                } else {
+                    return Some(FaultKind::AgreementEpoch);
+                }
+            }
+        }
+        None
+    }
+
+    /// Creates message content from `ReceivedMessages`. That message content, together with a
+    /// target node ID, forms messages that can be sent to the target node.
```

Comment:
That comment is a bit confusing. It reads as if it were an outgoing queue. Maybe:
"~`together with a target node ID`~"
"`sent to the target node`" → "`handled`"

Reply:
It's confusing indeed! Maybe we can reduce the last sentence even further to avoid speaking about conversion from `MessageContent` to `Message`? For example: "That message content can then be handled."